File: specodex/admin/blacklist.py
"""Manufacturer blacklist — manufacturers banned from promotion to prod.

The blacklist is a JSON file at repo root (``admin/blacklist.json``).
It is checked into git so changes go through code review.

Comparison is case-insensitive: "ABB", "abb", and "Abb" all refer to the
same entry, and the first-added casing is preserved for display. This
prevents trivial bypass and matches the TypeScript implementation in
``app/backend/src/services/blacklist.ts``.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Dict, List

DEFAULT_BLACKLIST_PATH = (
    Path(__file__).resolve().parents[2] / "admin" / "blacklist.json"
)


class Blacklist:
    """A set of manufacturer names banned from promotion to prod."""

    def __init__(self, path: Path = DEFAULT_BLACKLIST_PATH) -> None:
        self.path = path
        # Map from lowercased key → original casing. First-added wins.
        self._banned: Dict[str, str] = {}
        if path.exists():
            self.load()

    def load(self) -> None:
        data = json.loads(self.path.read_text())
        banned = data.get("banned_manufacturers", [])
        if not isinstance(banned, list):
            raise ValueError(
                f"{self.path}: 'banned_manufacturers' must be a list of strings"
            )
        self._banned = {}
        for raw in banned:
            name = str(raw)
            key = name.lower()
            # Preserve the first occurrence's casing if the file has duplicates.
            if key not in self._banned:
                self._banned[key] = name

    def save(self) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        payload = {"banned_manufacturers": self.names()}
        self.path.write_text(json.dumps(payload, indent=2) + "\n")

    def add(self, name: str) -> bool:
        """Add a manufacturer. Returns True if newly added, False if already present (case-insensitive)."""
        key = name.lower()
        if key in self._banned:
            return False
        self._banned[key] = name
        return True

    def remove(self, name: str) -> bool:
        """Remove a manufacturer. Returns True if removed, False if not present (case-insensitive)."""
        key = name.lower()
        if key not in self._banned:
            return False
        del self._banned[key]
        return True

    def contains(self, name: str) -> bool:
        """Case-insensitive membership check."""
        return name.lower() in self._banned

    def names(self) -> List[str]:
        """Stored (display) names, sorted case-insensitively."""
        return sorted(self._banned.values(), key=str.lower)

    def __len__(self) -> int:
        return len(self._banned)

File: app/backend/src/services/blacklist.ts
/**
 * Manufacturer blacklist — manufacturers banned from promotion to prod.
 *
 * Reads/writes the same JSON file the Python CLI uses
 * (specodex/admin/blacklist.py), so edits from either tool stay in sync.
 * File lives at repo root: admin/blacklist.json
 *
 * Comparison is case-insensitive: "ABB", "abb", and "Abb" all refer to the
 * same entry, and the first-added casing is preserved for display. This
 * prevents trivial bypass and matches the Python side.
 */

import fs from 'fs';
import path from 'path';

// Resolve to <repo_root>/admin/blacklist.json. This file lives at
// app/backend/src/services/blacklist.ts — four parents up is the repo root.
// Matches the pattern used by config/index.ts for loading .env.
const BLACKLIST_PATH = path.resolve(__dirname, '../../../..', 'admin', 'blacklist.json');

interface BlacklistFile {
  banned_manufacturers: string[];
}

export class Blacklist {
  private readonly filePath: string;
  // Map from lowercased name → original casing. First-added wins on collision.
  private banned: Map<string, string>;

  constructor(filePath: string = BLACKLIST_PATH) {
    this.filePath = filePath;
    this.banned = new Map();
    this.load();
  }

  load(): void {
    if (!fs.existsSync(this.filePath)) {
      this.banned = new Map();
      return;
    }
    const raw = fs.readFileSync(this.filePath, 'utf-8');
    const parsed = JSON.parse(raw) as BlacklistFile;
    if (!Array.isArray(parsed.banned_manufacturers)) {
      throw new Error(
        `${this.filePath}: 'banned_manufacturers' must be an array of strings`
      );
    }
    this.banned = new Map();
    for (const rawName of parsed.banned_manufacturers) {
      const name = String(rawName);
      const key = name.toLowerCase();
      // Preserve first occurrence's casing if the file has duplicates.
      if (!this.banned.has(key)) this.banned.set(key, name);
    }
  }

  save(): void {
    const dir = path.dirname(this.filePath);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
    const payload: BlacklistFile = {
      banned_manufacturers: this.names(),
    };
    fs.writeFileSync(this.filePath, JSON.stringify(payload, null, 2) + '\n');
  }

  /** Returns true if newly added, false if already present (case-insensitive). */
  add(name: string): boolean {
    const key = name.toLowerCase();
    if (this.banned.has(key)) return false;
    this.banned.set(key, name);
    return true;
  }

  /** Returns true if removed, false if not present (case-insensitive). */
  remove(name: string): boolean {
    const key = name.toLowerCase();
    if (!this.banned.has(key)) return false;
    this.banned.delete(key);
    return true;
  }

  /** Case-insensitive membership check. */
  contains(name: string): boolean {
    return this.banned.has(name.toLowerCase());
  }

  /** Returns the stored (display) names, sorted case-insensitively. */
  names(): string[] {
    return [...this.banned.values()].sort((a, b) =>
      a.toLowerCase().localeCompare(b.toLowerCase())
    );
  }

  get size(): number {
    return this.banned.size;
  }
}

export const BLACKLIST_FILE_PATH = BLACKLIST_PATH;
